src/Scale.hs

```diff
@@ -1,99 +1,199 @@
+baseModeIntervals HarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Major, Major, Perfect, Perfect, Minor, Major] [1..7]
+baseModeIntervals DoubleHarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Minor, Major, Perfect, Perfect, Minor, Major] [1..7]
+
+
+modeToIntervals :: Mode -> Set Interval
+modeToIntervals (Mode baseMode exts) = foldr extIntervals (baseModeIntervals baseMode) exts
+  where 
+    extIntervals :: ScaleExt -> Set Interval -> Set Interval
+    extIntervals ext intSet = insert (oldInt <+> (impliedShift $ acc ext)) (delete oldInt intSet)
+      where
+        -- TODO: If there isn't only one interval of a certain degree, the mode is
+        -- ambiguously constructed and we should give a warning.
+        oldInt = elemAt 0 (S.filter (\a -> getSize a == deg ext) intSet)    
+
+
+scaleToNotes :: Scale -> Set Root
+scaleToNotes (Scale root mode) = mapMonotonic (flip jumpIntervalFromNote root) (modeToIntervals mode)
+
+
+modalDistance :: Set Interval -> Set Interval -> Int
+modalDistance mode1 mode2 = sum $ intDistance <$> zip (toAscList mode1) (toAscList mode2)
+  where 
+    intDistance :: (Interval, Interval) -> Int
+    intDistance (i1, i2) = abs $ fromJust $ intervalToDistance (i1 |-| i2) 
+
+
+modesToExts :: Set Interval -> Set Interval -> [ScaleExt]
+modesToExts mode1 mode2 = 
+  let 
+    zippedInts = zip (toAscList mode1) (toAscList mode2)  
+    intervalDiffToAcc :: Interval -> Interval -> Accidental
+    intervalDiffToAcc i1 i2 = shiftToAcc $ fromJust $ intervalToDistance $ i2 |-| i1
+    accToExtList :: Accidental -> Int -> [ScaleExt] -> [ScaleExt]
+    accToExtList accidental degree = 
+      if accidental == natural then
+        id
+      else  
+        (++ [ScaleExt { acc = accidental
+                      , deg = degree 
+                      }])
+  in  
+    foldr (\(i1,i2) exts -> accToExtList (intervalDiffToAcc i2 i1) (getSize i1) exts) 
+          []
+          zippedInts 
+
+
+intervalsToMode :: Set Interval -> [Mode]
+intervalsToMode intSet = 
+  let 
+    sameDegreeModes = filter (\a -> S.map getSize (baseModeIntervals a) 
+                                 == S.map getSize intSet) 
+                      [Lydian ..]
+    distanceFromIntSet :: Set Interval -> BaseMode -> Int
+    distanceFromIntSet iSet mode = modalDistance iSet $ baseModeIntervals mode
+    sortedModes = sortBy (compare `on` distanceFromIntSet intSet) sameDegreeModes
+    exts = modesToExts intSet <$> (baseModeIntervals <$> sortedModes)
+  in
+    takeWhile (\mode -> numAlteredDegsInMode mode == minimum (length <$> exts)) $ (\(x,y) -> Mode x y) <$> zip sortedModes exts
+
+
+isSubsetMode :: Set Interval -> Set Interval -> Bool
+isSubsetMode mode1 mode2 = isSubsetOf mode1 mode2 
+
+getSubsetModeByDegree :: Set Interval -> Set Int -> Set Interval
+getSubsetModeByDegree mode degs = S.filter (\i -> any (== getSize i) degs) mode
+
+invert :: Set Interval -> Set Interval
+invert mode = S.map I.invert mode
```

Comment:
This all makes sense, though the invert for the chord will be tough, because it's mostly about voicing and we haven't gotten there yet. Still think we should go for it though to avoid new names for everything.

Reply:
Oh yeah, what's great about this approach is that you can write the instances when you're ready - the class itself can exist way before any of its instances.